ddl/partition.go

```diff
@@ -1184,14 +1185,31 @@ func buildRangePartitionDefinitions(ctx sessionctx.Context, defs []*ast.Partitio
 
 		buf := new(bytes.Buffer)
 		// Range columns partitions support multi-column partitions.
-		for _, expr := range clause.Exprs {
+		for i, expr := range clause.Exprs {
 			expr.Accept(exprChecker)
 			if exprChecker.err != nil {
 				return nil, exprChecker.err
 			}
-			expr.Format(buf)
-			piDef.LessThan = append(piDef.LessThan, buf.String())
-			buf.Reset()
+			// If multi-column use new evaluated+normalized output, instead of just formatted expression
```

Comment:
I still do not understand why `multi-column use new evaluated+normalized output`?  look like it is not compatibility with mysql:

```sql
create table t(a time, b time) partition by range columns (a, b) (partition p1 values less than ('2020', '2020'), partition p2 values less than ('20:20:10', '20:20:10'));
```

then, `show create table` result in MySQL 8.0 is:
```sql
mysql>show create table t\G
***************************[ 1. row ]***************************
Table        | t
Create Table | CREATE TABLE `t` (
  `a` time DEFAULT NULL,
  `b` time DEFAULT NULL
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_0900_ai_ci
/*!50500 PARTITION BY RANGE  COLUMNS(a,b)
(PARTITION p1 VALUES LESS THAN ('2020','2020') ENGINE = InnoDB,
 PARTITION p2 VALUES LESS THAN ('20:20:10','20:20:10') ENGINE = InnoDB) */
``` 

As you can see, MySQL keep use the original value in `PARTITION p1 VALUES LESS THAN ('2020','2020')`

The result in TiDB with this PR is:
```sql
mysql>show create table t\G
***************************[ 1. row ]***************************
Table        | t
Create Table | CREATE TABLE `t` (
  `a` time DEFAULT NULL,
  `b` time DEFAULT NULL
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin
PARTITION BY RANGE COLUMNS(`a`,`b`)
(PARTITION `p1` VALUES LESS THAN ('00:20:20','00:20:20'),
 PARTITION `p2` VALUES LESS THAN ('20:20:10','20:20:10'))
```

But TiDB will use a converted value in `PARTITION `p1` VALUES LESS THAN ('00:20:20','00:20:20')`

Reply:
It is to make it more user friendly. The output is still compatible with MySQL, meaning the output is still valid syntax, just more readable. I did not change it for the existing single column RANGE COLUMNS, but only for this new implementation, to avoid having to do it in a new feature request. Also it replaces the `expression.Format()` call on the original expression, to a `value.String()` from the validation step, where the expression is already evaluated and checked.

In your example it is much more clear that '00:20:00' is a time value, rather than '2020' which needs the context of the column type as time and some mental parsing and mysql knowledge to understand that it means the same.

I see this more of a nice to have, which makes the feature better (better readability). It is not necessary from a technical point.

@crazycs520 do you want me to remove this to make the output the same as MySQL or do you see any value of having the partition definition values evaluated and normalized?

Reply:
In my opinion, line #1193 ~ #1208 make the logic here more complicated, and the complicated is not necessary.
For compatibility or make it more user-friendly, how about asking for PM's opinion first?

Reply:
@tiancaiamao PTAL

Reply:
After discussing with @bb7133 and PM @moyun, they prefer the currently user-friendly implementation, I agree too.